### ISSUE TITLE
Usage of resource not detected

### ISSUE THREAD

**Issue by bbokorney** [concourse/atc#132]
# Bug Report

* Concourse version: 2.3.1
* Deployment type: BOSH
* Browser: N/A

Version 2.3.0 of Concourse introduced the a validation error for unused resource (in [this PR](https://github.com/concourse/atc/pull/116)). There appears to be a scenario where the validation fails to detect that a resource is being used. Specifically, resources in an `on_success`, `on_failure`, or `ensure` block attached to an `aggregate` block are not detected. Here's an example pipeline configuration which demonstrates the defect.

```yaml
resources:
- name: slack-notification
  type: slack
  source:
    url: {{slack-url}}

resource_types:
- name: slack
  type: docker-image
  source:
    repository: cfcommunity/slack-notification-resource
    tag: latest

jobs:
- name: do-it
  plan:
  - aggregate:
    - task: this-will-fail
      config:
        platform: linux
        image_resource:
          type: docker-image
          source:
            repository: alpine
            tag: latest
        run:
          path: cat
          args: [not-a-file.txt]
    on_success:
      put: slack-notification
      params:
        text: Task failed
    on_failure:
      put: slack-notification
      params:
        text: Task succeeded
    ensure:
      put: slack-notification
      params:
        text: This should always run

```

When setting this pipeline, I get the following error:

```bash
$ fly -t test set-pipeline -c pipeline.yml -p test-pipeline -l creds.yml
...
error: invalid configuration:
invalid resources:
	resource 'slack-notification' is not used
```

In the above example, the `slack-notification` resource is used only in the three blocks attached to the `aggregate`. Interestingly, if an aggregate is _not_ used (if the`on_success`, `on_failure`, or `ensure` are attached directly to a task), the validation correctly recognizes the `slack-notification` resource is used.

**Comment by vito** [comment]
Can you check if this is fixed in 2.4.0? 'Cause I think it is.

**Comment by bbokorney** [comment]
I ran 2.4.0 locally and I'm able to set this pipeline and it works as expected. I'll have to upgrade our Concourse deployment. Thanks for the quick response!
